src/services/fileStorage/strategies/awsS3.js

```diff
@@ -1,46 +1,102 @@
 const { promisify } = require('es6-promisify');
+const CryptoJS = require('crypto-js');
 const { BadRequest, NotFound, GeneralError } = require('@feathersjs/errors');
 const aws = require('aws-sdk');
 const pathUtil = require('path');
 const fs = require('fs');
 const logger = require('../../../logger');
 const { schoolModel } = require('../../school/model');
+const storageProviderModel = require('../../storageProvider/model');
 const UserModel = require('../../user/model');
 const filePermissionHelper = require('../utils/filePermissionHelper');
 const { removeLeadingSlash } = require('../utils/filePathHelper');
 
-// const prodMode = process.env.NODE_ENV === 'production';
+const AbstractFileStorageStrategy = require('./interface.js');
+
+const getCorsRules = () => ([{
+	AllowedHeaders: ['*'],
+	AllowedMethods: ['PUT'],
+	AllowedOrigins: [process.env.HOST],
+	MaxAgeSeconds: 300,
+}]);
+
+const getConfig = (provider) => {
+	const awsConfig = new aws.Config({
+		signatureVersion: 'v4',
+		s3ForcePathStyle: true,
+		sslEnabled: true,
+		accessKeyId: provider.accessKeyId,
+		secretAccessKey: provider.secretAccessKey,
+		region: provider.region || 'eu-de',
+		endpointUrl: provider.endpointUrl,
+		cors_rules: getCorsRules(),
+	});
+	awsConfig.endpoint = new aws.Endpoint(provider.endpointUrl);
+	return awsConfig;
+};
+
+const chooseProvider = async (schoolId) => {
+	const provider = await storageProviderModel.aggregate([
+		{ $match: { isShared: true } },
+		{ $addFields: { freeBuckets: { $subtract: ['$maxBuckets', { $size: '$schools' }] } } },
+		{ $sort: { freeBuckets: -1 } },
+		{ $limit: 1 },
+	]).exec();
```

Comment:
Eventuell wäre es billiger statt maxBuckets und schools gleich auch freeBuckets im Model zu haben, dann wäre das hier keine Aggregation. Muss aber vermutlich nicht sein, da das nur bei der Erstellung der Schule gerufen wird, korrekt?

Reply:
Jap, das wird pro Schule nur einmal in der Lifetime aufgerufen, von daher lieber so lassen :)